src/main/java/com/google/sps/servlets/GenerateServlet.java

```diff
@@ -28,14 +30,171 @@
 
   private static final String FILTER_ARRAY = "clicked-array";
 
+  String[] allPlaces = {"Paris", "New York City", "San Francisco", "London", "Sydney", "Venice"};
+  String[] allDifficulties = {"Easy", "Medium", "Hard"};
+
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    HashSet<String> userPlaces = new HashSet();
+    HashSet<String> userDifficulties = new HashSet();
+
+    ArrayList<Destination> allDestinations = new ArrayList();
+    // ArrayList<Name of Destination>
+    ArrayList<String> filteredDestinations = new ArrayList();
+
     // Get array of clicked filters and convert to ArrayList<String>
-    Gson gson = new Gson();
     HashSet<String> clickedFilters =
-        gson.fromJson(request.getParameter(FILTER_ARRAY), HashSet.class);
+        new Gson().fromJson(request.getParameter(FILTER_ARRAY), HashSet.class);
+
+    allDestinations = createFakeDestinations();
+    userPlaces = separatePlaceFilters(clickedFilters);
+    userDifficulties = separateDifficultyFilters(clickedFilters);
+    filteredDestinations = filterPlaces(userPlaces, allDestinations);
+    filteredDestinations =
+        filterDifficulty(filteredDestinations, userDifficulties, allDestinations);
+    writeToDataStore(filteredDestinations);
 
     response.setContentType("text/html;");
     response.getWriter().println(clickedFilters);
   }
+
+  /* Separate place filters from rest of user filters. */
+  public HashSet<String> separatePlaceFilters(HashSet<String> clickedFilters) {
+    HashSet<String> userPlaces = new HashSet();
+    for (int i = 0; i < allPlaces.length; i++) {
+      if (clickedFilters.contains(allPlaces[i])) {
+        userPlaces.add(allPlaces[i]);
+      }
+    }
+    return userPlaces;
+  }
+
+  /* Return HashSet of only difficulty filters. */
+  public HashSet<String> separateDifficultyFilters(HashSet<String> clickedFilters) {
+    HashSet<String> userDifficulties = new HashSet();
+    for (int i = 0; i < allDifficulties.length; i++) {
+      if (clickedFilters.contains(allDifficulties[i])) {
+        userDifficulties.add(allDifficulties[i]);
+      }
+    }
+    return userDifficulties;
+  }
+
+  /* Return Destination objects within specified place. */
+  public ArrayList<String> filterPlaces(
+      HashSet<String> userPlaces, ArrayList<Destination> allDestinations) {
+    ArrayList<String> filteredDestinations = new ArrayList();
+    // Iterate through destinations
+    // If Destination place is in userPlaces array
+    // Add Destination to filteredDestinations array
+    for (int i = 0; i < allDestinations.size(); i++) {
+      Destination currDestination = allDestinations.get(i);
+      if (userPlaces.contains(currDestination.getCity())) {
```

Comment:
```suggestion
      if (userPlaces.contains(destination.getCity())) {
```

Reply:
so you can do this instead of creating a new variable to assign the current destination to